crates/lib3h/src/transport/transport_multiplex/mod.rs

```diff
@@ -0,0 +1,225 @@
+                TransportMultiplex::new(Box::new(TransportMock::new(s_out, r_in))),
+                "test_mplex_",
+            );
+
+        let mut route_a: TransportActorParentWrapperDyn<()> = GhostParentWrapperDyn::new(
+            Box::new(
+                mplex
+                    .as_mut()
+                    .get_agent_space_route("space_a".to_string(), "agent_a".to_string()),
+            ),
+            "test_route_a_",
+        );
+
+        let mut route_b: TransportActorParentWrapperDyn<()> = GhostParentWrapperDyn::new(
+            Box::new(
+                mplex
+                    .as_mut()
+                    .get_agent_space_route("space_b".to_string(), "agent_b".to_string()),
+            ),
+            "test_route_b_",
+        );
+
+        // send a message from route A
+        route_a.request(
+            std::time::Duration::from_millis(2000),
+            (),
+            RequestToChild::SendMessage {
+                address: addr_none.clone(),
+                payload: b"hello-from-a".to_vec(),
+            },
+            Box::new(|_, _, response| {
+                assert_eq!(&format!("{:?}", response), "");
+                Ok(())
+            }),
+        );
+
+        route_a.process(&mut ()).unwrap();
+        mplex.process(&mut ()).unwrap();
+        route_a.process(&mut ()).unwrap();
+        mplex.process(&mut ()).unwrap();
+
+        // should receive that out the bottom
+        let (address, payload) = r_out.recv().unwrap();
+        assert_eq!(&addr_none, &address);
+        assert_eq!(&b"hello-from-a".to_vec(), &payload);
+
+        // send a message up the bottom
+        s_in.send((addr_none.clone(), b"hello-to-b".to_vec()))
+            .unwrap();
+
+        // process "receive" that message
+        mplex.process(&mut ()).unwrap();
+        let mut msgs = mplex.drain_messages();
+        assert_eq!(1, msgs.len());
+
+        let msg = msgs.remove(0).take_message().unwrap();
+        if let RequestToParent::ReceivedData { address, payload } = msg {
+            assert_eq!(&addr_none, &address);
+            assert_eq!(&b"hello-to-b".to_vec(), &payload);
+        } else {
+            panic!("bad type");
+        }
+
+        // our mplex module got it, now we should have the context
+        // let's instruct it to be forwarded up the route
+        mplex.as_mut().received_data_for_agent_space_route(
+            "space_b".to_string(),
+            "agent_b".to_string(),
+            "agent_x".to_string(),
+            "machine_x".to_string(),
+            b"hello".to_vec(),
+        );
+
+        mplex.process(&mut ()).unwrap();
+        route_b.process(&mut ()).unwrap();
+
+        let mut msgs = route_b.drain_messages();
+        assert_eq!(1, msgs.len());
+
+        let msg = msgs.remove(0).take_message().unwrap();
+        if let RequestToParent::ReceivedData { address, payload } = msg {
+            assert_eq!(
+                &Url::parse("transportid:machine_x?a=agent_x").unwrap(),
```

Comment:
Is this a canonical url representation or just thrown together for the test?

Reply:
this is canonical / correct test